File: python/heap-priority-queue/last-stone-weight.py
import heapq

class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        heap = []
        for i in stones:
            heapq.heappush(heap, -i)

        while len(heap) >= 2:
            y = -heapq.heappop(heap)
            x = -heapq.heappop(heap)
            if x == y:
                continue
            else:
                heapq.heappush(heap, -(y - x))

        if heap:
            return -heapq.heappop(heap)
        else:
            return 0

File: c++/heap-priority-queue/last-stone-weight.cc
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        std::priority_queue<int> pq;

        for (int stone : stones) {
            pq.push(stone);
        }

        while (pq.size() > 1) {
            int x = pq.top();
            pq.pop();
            int y = pq.top();
            pq.pop();

            if (x == y) {
                continue;
            } else {
                pq.push(x - y);
            }
        }

        if (pq.size() == 1) {
            return pq.top();
        } else {
            return 0;
        }
    }
};
